### PR TITLE
Fix certificate LocalDate parsing being too lenient (DEV)

### PR DESCRIPTION
**Pull request by d4rken**
While #3418 made parsing a bit more lenient to accept Polish certificates, it's now a bit too lenient :facepalm:.
Missing month or day elements were assumed to be "start" (e.g. 2021 -> 2021-01-01).
While no such QR codes should be officially generated, better safe than sorry.

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/common/certificate/Dcc.kt [hunk 1]
```diff
@@ -1,9 +1,14 @@
...
 import com.google.gson.annotations.SerializedName
+import org.joda.time.DateTime
 import org.joda.time.LocalDate
+import org.joda.time.format.DateTimeFormat
+import org.joda.time.format.DateTimeFormatterBuilder
+import org.joda.time.format.ISODateTimeFormat
+import timber.log.Timber
 
-interface Dcc<PayloadType : Dcc.Payload> {
+abstract class Dcc<PayloadType : Dcc.Payload> {
```
Comment (chris-cwa) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/common/certificate/Dcc.kt:11]: I'm just curious: What was wrong with being an interface?

Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/common/certificate/Dcc.kt:11]: Interface doesn't allow backing fields, by making it an abstract class I can move the caching for `dateOfBirth` parsing into the `Dcc` class and reuse it for all types of certificates.

**Comment by vaubaehn** [comment]
I found an updated schema here: https://github.com/ehn-dcc-development/ehn-dcc-schema
Looks like all related IT will need to adjust. Parsing with regex could be good to catch all future "adjustments".

**Review by mtwalli — APPROVED** [review]
Lgtm

**Review by BMItr — APPROVED** [review]
👌 safe

**Comment by vaubaehn** [comment]
@d4rken 
> Missing month or day elements were assumed to be "start" (e.g. 2021 -> 2021-01-01).
> While no such QR codes should be officially generated, better safe than sorry.

As you can see here: https://github.com/ehn-dcc-development/ehn-dcc-schema/issues/78 , such QR codes are already issued, and NL won't probably the only country.

FYI: @mlenkeit

**Pull request merged**